### ISSUE TITLE
Feature request: Linux: Improving Command display/sort

### ISSUE THREAD

**Issue by gnarendran** [hishamhm/htop#801]
OS: Linux

htop relies on /proc/pid/cmdline for populating the Command column. But this at times causes confusion, as any process is able to modify its own cmdline by writing into argv, by setproctitle() etc (for example chrome, GNU screen etc. do this). Ideally, program path/basename should be obtained by readlink of /proc/pid/exe, comm by reading /proc/pid/comm, and cmdline from /proc/pid/cmdline (unprocessed). Though only root is able to readlink ALL /proc/pid/exe, even for non-root users it is still an improvement to use it where permission allows, and resort to cmdline otherwise. And for root, it is a big improvement to be able to see the unobfuscated path for all processes.

The attached patch 
[htop_patch.txt](https://github.com/hishamhm/htop/files/2104898/htop_patch.txt), with a screenshot of the patched htop 
![htop_patched](https://user-images.githubusercontent.com/23692893/41451928-fcab62f4-708c-11e8-83a4-62e42d46c199.png)

features the following:

1. Command displays "exe|comm|cmdline", with comm displayed in magenta. For example with 'Show program path':

while the current htop displays (even when run as root):
 2293 root       20   0  399M  277M  1048 S  0.0  3.5  0:14.08 SCREEN -dAR

the patched htop displays ( │screen│ is in magenta (SOFTIRQ color)):
 2293 root       20   0  399M  277M  1048 S  0.2  3.5  0:14.14 /usr/bin/screen│screen│SCREEN -dAR

2. Display unprocessed /proc/pid/cmdline at the end. This helps to know the current state of argv, and solve issues like https://github.com/hishamhm/htop/issues/586. For example without 'Show program path':

while the current htop displays:
 2732 user  20   0  590M 88696 23880 S  0.0  1.1  0:01.95 07 --log-level=3 --service-request-channel-token=D8605083D929EE1C2FBA0586CCE0B639

the patched htop displays ( │chrome│ is in magenta):
 2732 user  20   0  590M 88696 23880 S  0.0  1.1  0:01.96 chrome│chrome│/opt/google/chrome/chrome --type=gpu-process --field-trial-handle=5678600059915485073,9407740234176539478,131072 --log-level=3 --service-request-channel-token=D8605083D929EE1C2FBA0586CCE0B639

3. Display /proc/pid/comm in the middle. This helps to know the name the user invoked the command with (could be different from exe for example if symlink is used or used for thread name.) Separate columns not introduced for exe and comm, as these are variable length fields and the resulting indentation would not be pleasing. Instead, used separators and color for comm, to get fair readability.

4. Sorting on Command is based on the displayed string, whether it is the full path or the basename that is displayed. This is an improvement over the current htop which sorts always on the comm, even when only the basename is displayed, which could be confusing.

5. No other column is affected, and other functionalities of the Command column are retained (tree view, custom thread names, basename highlighting etc.,)

I am aware these changes could be debatable/improvable, but please consider these changes at least enabled via settings.

**Comment by Explorer09** [comment]
I wonder if you can convert the patch file into a pull request? Thanks for the proposal, by the way.

**Comment by Explorer09** [comment]
@gnarendran 
I watched your screenshot for a while, and have a concern about the proposal, that the separate "exe|comm|cmdline" display looks too verbose, especially when it's common to see same basenames in all three fields.
I would propose reverting to older one-field display if all three happen to have the same content.
I was thinking of other ways to condense the display, though.

**Comment by Explorer09** [comment]
@gnarendran
I would propose this logic:
1. If _comm_ value matches the basename of _cmdline[0]_ (match the first (TASK_COMM_LEN-1) characters only; you know why), then hide the _comm_ value from display.
2. If _comm_ value matches the basename of _cmdline[1]_, then hide the _comm_ value, but highlight the basename of _cmdline[1]_. (This case can happen on scripts running on an interpreter, for example the bash & python script processes on your screenshot.)
3. Regardless of whether the _comm_ value is displayed, if the path of _exe_ matches _cmdline[0]_ in full, merge _exe_ and _cmdline[0]_ into one display.
4. If _exe_ only matches the basename of _cmdline[0]_, and _cmdline[0]_ contains no slashes (i.e. just a simple command name like `bash` or `ls` where its full _exe_ path was resolved via $PATH variable), merge _exe_ and _cmdline_ fields together where _cmdline[0]_ is replaced with the path of _exe_.
5. If _exe_ and _cmdline_ don't match in either of the methods above (this is rare, but it can happen on (a.) Google Chrome case, where the process messes up its own _cmdline_, or (b.) when both symlink and $PATH are used in resolving _exe_ path, e.g. "python3" resolves to "/usr/bin/python3" via $PATH and "/usr/bin/python3" is a symlink to "/usr/bin/python3.6"), show both fields separately.

For the display, I suggest instead of "exe|comm|cmdline", show in this format: "[comm | exe] cmdline"
This way we can have some compatibility with `top`'s format.

**Comment by gnarendran** [comment]
@Explorer09 
It appears we disagree on a few things, as follows:
1. I agree that for matching the comm prefix, using TASK_COMM_LEN (defined based on man proc(5)) is best - I have updated the patch accordingly. But I think it is better to match comm, as in the patch, with basename of exe (rather than basename of cmdline[0]), as cmdline is arbitrarily modifiable and in fact it is hard to find its basename reliably - see for example discussion in https://github.com/hishamhm/htop/issues/586
2. Finding if comm is the basename of an interpreted script, is even harder to do reliably. Because cmdline[1] need not be the script but might be an option (for example "/bin/bash -p bin/sysmon.sh"). There could be any number of options preceding the script in fact, and again locating its basename is tricky as in 1.
3. Thought about merging exe and cmdline even if |comm| is displayed, but the result would not be pleasing. This is because |comm| would appear after (say at the far end of a long) unified cmdline in some cases, and before the non-unified cmdline in some other cases, and so would seem to be  jerky to the user.
4. This was already done in the patch, where I mentioned "check if exe or its basename is a prefix of cmdline - if so strip that prefix and display the rest of cmdline in one field" - screenshot has the examples 2430(screen) and 4687(tail) where exe and cmdline have been unified.
5. This was again already done in the patch, screenshot has the example 2862(python3.6)
6. For display, I still like "exe|comm|cmdline" - the reason being, the user would intuitively expect the string being sorted on (whether full path or basename, whether based on exe (if available) or cmdline) to appear at the start.

But I guess such disagreements are normal as we are discussing UI and tastes differ. So we could keep this issue open for future thoughts and not do anything about it now. I am personally satisfied with the resulting patch as the modified htop displays the reliable (and unobfuscatable) exe and uses its full path/basename for sorting etc., and redundant information is mostly removed.

**Comment by Explorer09** [comment]
@gnarendran 

> 1. [...] But I think it is better to match comm, as in the patch, with basename of exe (rather than basename of cmdline[0]), as cmdline is arbitrarily modifiable and in fact it is hard to find its basename reliably - see for example discussion in #586

According to [man page](http://man7.org/linux/man-pages/man5/proc.5.html), comm is modifiable. If I am to rate the stability of the three fields, it would be like this: "comm < cmdline < exe". cmdline may be changed after an execve() system call, while exe path changes only when the file is moved on the filesystem. Since comm may be modified by the thread at will, I see no sense to compare it to exe path in the logic.

> 2. Finding if comm is the basename of an interpreted script, is even harder to do reliably. Because cmdline[1] need not be the script but might be an option (for example "/bin/bash -p bin/sysmon.sh"). There could be any number of options preceding the script in fact, and again locating its basename is tricky as in 1.

The goal was simply to avoid repeating display on htop "command" field since the screen space is limited.
You can modify my logic so that it can match cmdline[2], cmdline[3] and further until the text to be highlighted would go off the screen. How will this sound, then?

> 3. Thought about merging exe and cmdline even if |comm| is displayed, but the result would not be pleasing. This is because |comm| would appear after (say at the far end of a long) unified cmdline in some cases, and before the non-unified cmdline in some other cases, and so would seem to be jerky to the user.

If the user have the option to hide comm altogether, that won't be a problem.

> 6. For display, I still like "exe|comm|cmdline" - the reason being, the user would intuitively expect the string being sorted on (whether full path or basename, whether based on exe (if available) or cmdline) to appear at the start.

I have not concerned about sort key too much. My reason was to front-load differentiating keywords -- [something you'll do when writing content for online](http://www.kendallcopywriting.co.uk/front-loading-important-copywriting-online/) or with SEO (search engine optimization) in mind. And I was concerning about tree view -- it's better to see thread names rather then all-the-same exe prefixes in a process subtree.

**Comment by gnarendran** [comment]
@Explorer09 

1. Of course the three are modifiable, cmdline by the program arbitrarily, comm by threads/symlink invocation etc., and exe by the kernel. But we are just trying to condense the information, so by hiding comm when it matches with exe, we are simply saying to the user, you will find it in the basename of exe, otherwise will simply display comm so there is no issue. We could try to match it with the cmdline instead as well, only it is harder to find its basename reliably (again the open issue https://github.com/hishamhm/htop/issues/586). It doesn't matter otherwise where we tell the user to look for the hidden comm.

2. I do understand we can heuristically try to look for comm in cmdline. The problem is how reliably can we do it with cmdline. For example, comm might match a first genuine (string) argument and be mistaken for an interpreted script and so hidden, when probably it should've been displayed. With exe there is no such confusion, if it differs from basename display it, otherwise hide it and exe's basename has it.

3. I agree the user should have an option whether to display comm or not. Someone taking this forward could consider it indeed.

6. Though it is not clear to me what exactly your use case is, as I had stated in the first comment, thread/tree display is not particularly affected in the patch and it still provides  "Custom Thread Names" when opted for etc. The screenshot has examples. And of course, it is good to know the exe from a security standpoint, and nice to have the sorting on Command be intuitive (based on what is displayed) etc., which are my motivations.

**Comment by Explorer09** [comment]
@gnarendran
I think for argument's sake let's ignore #586 for now. If Google Chrome gave a misleading cmdline[0] for us to parse, we could point that out in htop's side, as long as we show the correct exe path.

I can't imagine a case that comm matches the basename of exe path but does not march any argument of cmdline, though.

2. When I propose to hide the comm value, I said the value should be highlighted within the cmdline argument. So technically no information would be lost, although there is a change of a wrong token being highlighted. If that happens, so be it.

Example: /bin/bash /home/explorer/my-super-long-shell-script.sh my-super-long-string-input
I am fine if it highlights like this: /bin/**bash** /home/explorer/**my-super-long-s**hell-script.sh my-super-long-string-input
Despite the comm might come from the string-input thing.

**Comment by gnarendran** [comment]
1. It happens at times even in genuine cases - see for example the 2439(screen|SCREEN) in the screenshot. And a mischievous program could of course do it.
2. Effectively you are saying LinuxProcess_writeCommand should, after looking at the basename without slash, do a strstr for "/<comm>" on the concatenated cmdline, and highlight/colorize the hit. In most cases it would score a hit on the basename or the interpreted script. In very rare cases it would: a) score a hit on a unrelated string embedded (even far away) in the cmdline and you are ok with it; b) not score a hit at all, and only then optionally display comm as a separate field.

I agree this is a fair approach as well, accompanied by suitable documentation so that the user won't be surprised in the above rare cases.

**Comment by gnarendran** [comment]
2. A reasonable bad case for the heuristics will be when comm is "my-long-script-" and cmdline ends up being displayed "/bin/bash -p /home/user/**my-long-script-**name-examples-dir/my-long-script-name-example1.sh"

**Comment by Explorer09** [comment]
@gnarendran 
1. I would wonder why `screen` would produce an all-caps `SCREEN` program name in its cmdline[0], and whether's it's a common use case. Nevertheless, if you think you can reduce text duplication by comparing comm name with exe basename, then go ahead, although I would expect it's compared second.
2. 

> In very rare cases it would: a) score a hit on a unrelated string embedded (even far away) in the cmdline and you are ok with it; b) not score a hit at all, and only then optionally display comm as a separate field.

Yes, that was the behavior I expected. Because the highlight would be 15 (or (TASK_COMM_LEN-1)) characters in maximum, we can reason about the mis-highlighting. In the visual and in the manual.

But your example given is wrong:

> [W]hen comm is "myLongScriptNam" and cmdline ends up being displayed "/bin/bash -p /home/user/myLongScriptName-examples-dir/myLongScriptName-example1.sh"

Basename function will strip everything before the final slash, so it would be "**myLongScriptNam**e-example1.sh" part that would be highlighted. Except, of course, when the slash gets cut off due to screen space limit.

**Comment by gnarendran** [comment]
1. Historically many UNIX flavors (*BSD, Linux) have allowed cmdline to be used as a billboard for a process to advertise its current state, when viewed using ps, top etc.. But unfortunately there has not been a standard (in spite of recent setproctitle()) for how cmdlline should be modified. For example a shell writes a 'minus' when it is a login shell (e.g, -/bin/bash), a server instance of a process might say "process (server)" to differentiate itself from its client instance "process (client)", chrome seems to display all its dynamic options in argv[0] itself etc.. In the case of GNU screen which is a terminal multiplexer, it writes "SCREEN" to say it is the persistent instance that will stay even after detachment of the terminal, as differentiated from "screen" instance which appears when attached. Not justifying "SCREEN", but saying lack of standard has resulted in all these possibilities.

To sum up, comparing comm with: a) exe de-duplicates most cases (say 95% for comparison), but not symlink/interpreted scripts, and it is reliable b) cmdline de-duplicates even more cases (say 99%) including symlink/interpreted scripts, but in very rare cases might highlight unrelated string in cmdline, and the user should be made aware of it.

2. The example assumed using strstr "/comm" on the concatenated cmdline. But it seems you meant strtok(" ") followed by basename() followed by strncmp(), so the example does not apply, please ignore it.

**Comment by Explorer09** [comment]
@gnarendran Not strtok(" "), but using the built-in '\0' as the argument delimiter. This way it would work on argument with spaces in their file path.

**Comment by gnarendran** [comment]
@Explorer09 
But that would require re-reading /proc/pid/cmdline, as by the time of display cmdline has been already read and concatenated into a string (existing infrastructure.) And then there is the dreaded case of https://github.com/hishamhm/htop/issues/586 :-)

**Comment by gnarendran** [comment]
The above discussion is summed up in the patch 
[htop_patch-2.txt](https://github.com/hishamhm/htop/files/2133001/htop_patch-2.txt). The difference from the previous patch is that, after matching comm with basename of exe (reliable), a new display option "Try finding command name in cmdline" enables looking for comm in basename of tokens in cmdline (heuristic), and only after that fails (very rare) is comm displayed in a separate field. The entire token with the comm prefix is colored, as otherwise the strings look spotty and it hinders readability (a compiler switch in LinuxProcess_writeCommand, "COLORIZE_ONLY_COMM_PREFIX" changes this if needed). The assumption being the user knows only 15 bytes of could be comm. The screenshots without and with the new options enabled, are given below. Interesting rows are 1169(sd-pam), 2094(seapplet), 21879(test_symlinked). Hope this helps further the effort.
![htop_patched-2 0](https://user-images.githubusercontent.com/23692893/41848333-c2aeb702-789a-11e8-803e-b0da725036aa.png)
![htop_patched-2 1](https://user-images.githubusercontent.com/23692893/41848334-c2e1c2c8-789a-11e8-87a7-4f25268fe170.png)
![htop_patched-2 2](https://user-images.githubusercontent.com/23692893/41848335-c312aa1e-789a-11e8-9214-70381dd04a85.png)

**Comment by Explorer09** [comment]
@gnarendran
Time to make a pull request, wouldn't you?

1. Click on "Fork" on the top of this repository page.
2. Download and install "git".
3. Use git to make a branch on your repository, basically you run this:

```
git clone https://github.com/gnarendran/htop htop
cd htop
git checkout -b improved-command-display
(Apply your changes and)
git commit -a -m 'Improving Command display/sort'
git push origin improved-command-display
```

4. Make a pull request on GitHub by following [this](https://help.github.com/articles/creating-a-pull-request-from-a-fork/)

**Comment by gnarendran** [comment]
a) It is worth noting that readlink permission for /proc/pid/exe is governed by the capability CAP_SYS_PTRACE. See man pages proc(5), ptrace(2):PTRACE_MODE_READ_FSCREDS
So instead of running htop as root, it is a lot safer to only give this capability to htop during installation:
setcap 'cap_sys_ptrace=ep' /usr/bin/htop
This excludes cap_sys_nice and cap_kill that htop would otherwise get if run as root.

b) In the settings menu, it is probably more appropriate to place the new option 'Try to find command name in cmdline' after 'Highlight program "basename"'.
